lib/pbio/drv/reset/reset_ev3.c

```diff
@@ -7,33 +7,85 @@
 
 #if PBDRV_CONFIG_RESET_EV3
 
+#include <stdbool.h>
+
+#include <pbdrv/compiler.h>
 #include <pbdrv/reset.h>
 #include <pbdrv/gpio.h>
 
 #include "../drv/gpio/gpio_ev3.h"
 
+#include <tiam1808/hw/soc_AM1808.h>
 #include <tiam1808/hw/hw_syscfg0_AM1808.h>
+#include <tiam1808/hw/hw_types.h>
+#include <tiam1808/timer.h>
+
+#define BOOTLOADER_UPDATE_MODE_VALUE    0x5555AAAA
+
+// 'Pybr'
+#define RESET_REASON_FLAG_WDT           0x50796272
+// 'rboo'
+#define RESET_REASON_FLAG_SOFT_RESET    0x72626f6f
+
+typedef struct {
+    // 0xffff1ff0
+    uint32_t _dummy0;
+    // 0xffff1ff4
+    // Pybricks uses this flag to determine software reset vs other resets
+    uint32_t reset_reason_flag;
+    // 0xffff1ff8
+    // Have not fully investigated this, but the bootloader seems to store
+    // the result of (DDR) memory testing at 0xffff1ffa and 0xffff1ffb
+    uint32_t _bootloader_unk_ram_test;
+    // 0xffff1ffc
+    uint32_t bootloader_update_flag;
+} persistent_data_t;
+extern volatile persistent_data_t ev3_persistent_data;
+
+static bool second_init;
+static uint32_t saved_reset_reason_flag;
 
 static const pbdrv_gpio_t poweroff_pin = PBDRV_GPIO_EV3_PIN(13, 19, 16, 6, 11);
 
 void pbdrv_reset_init(void) {
     pbdrv_gpio_out_high(&poweroff_pin);
+    if (!second_init) {
+        saved_reset_reason_flag = ev3_persistent_data.reset_reason_flag;
+        // The EV3 happens to call this init function multiple times,
```

Comment:
We shouldn't be. Can we remove `pbdrv_reset_init()` from `SystemInit()` and replace it with something else (e.g. just put the power pin call inline there)?